x-pack/plugins/enterprise_search/KEA.md

```diff
@@ -0,0 +1,178 @@
+  },
+  reducers: () => ({
+    sourceApiStatus: [
+      {
+        status: 'IDLE',
+      },
+      {
+        fetchSource: () => ({
+          status: 'PENDING',
+        }),
+        fetchSourceError: (_, error) => ({
+          status: 'ERROR',
+          error,
+        }),
+        fetchSourceSuccess: (_, data) => ({
+          status: 'SUCCESS',
+          data,
+        }),
+      },
+    ],
+  }),
+  listeners: ({ actions }) => ({
+    fetchSource: async () => {
+      clearFlashMessages();
+
+      try {
+        const response = await HttpLogic.values.http.post<CustomSource>('/api/source');
+        actions.fetchSourceSuccess(response);
+      } catch (e) {
+        flashAPIErrors(e);
+        actions.fetchSourceError(e.code, e.message);
+      }
+    },
+  }),
+});
+```
+
+The types used above can be found in our [common Enterprise Search types file](common/types/api.ts). While the above assumes a single, idempotent API, this approach can be easily extended to use a dictionary approach:
+```
+reducers: () => ({
+    sourceApiStatus: [
+      {
+      },
+      {
+        fetchSource: (state, id) => ({...state,
+        id: {
+          status: 'PENDING',
+          data: state[id]?.data,
+        }}),
+        fetchSourceError: (_, ({id, error})) => ({...state,
+        id: {
+          status: 'ERROR',
+          error,
+        }}),
+        fetchSourceSuccess: (_, ({id, data})) => ({...state, id: {
+          status: 'SUCCESS',
+          data,
+        }}),
+      },
+    ],
+  }),
+```
+## Import actions and values from API logic files into component and view logic.
+
+Once you have an API interactions file set up, components and other Kea logic files can import the values from those files to build their own logic. Use the Kea 'connect' functionality to do this, as the auto-connect functionality has a few bugs and was removed in Kea 3.0. This allows you to read the status and value of an API, react to any API events, and abstract those APIs away from the components. Those components can now become more functional and reactive to the current state of the application, rather than to directly responding to API events.
+
+```
+export const AddCustomSourceLogic = kea<
+  MakeLogicType<AddCustomSourceValues, AddCustomSourceActions, AddCustomSourceProps>
+>({
+  connect: {
+    actions: [AddCustomSourceApiLogic, ['addSource', 'addSourceSuccess', 'addSourceError']],
+    values: [AddCustomSourceApiLogic, ['sourceApi']],
+  },
+  path: ['enterprise_search', 'workplace_search', 'add_custom_source_logic'],
+  actions: {
+    createContentSource: true,
+    setCustomSourceNameValue: (customSourceNameValue) => customSourceNameValue,
+    setNewCustomSource: (data) => data,
+  },
+  reducers: ({ props }) => ({
+    customSourceNameValue: [
+      props.initialValue || '',
+      {
+        setCustomSourceNameValue: (_, customSourceNameValue) => customSourceNameValue,
+      },
+    ],
+    newCustomSource: [
+      {} as CustomSource,
```

Comment:
I know this isn't the point of this example but in general I'm not a huge fan of using `{} as Y` in these logic files to coerce empty values into, I'd rather we set it as `null` if its empty to start.  I know it does save us from having to a bunch of `=== null` checks though.

Reply:
Is there an idiomatic way of creating empty initialization for TypeScript ? 

I can think of rust's `Default` trait as an example, which I believe could help us on here and many other places.

Reply:
There's not, no. TypeScript is a very functional rather than object-oriented language so things like traits don't really work.

I mostly agree with Byron here that the idiomatic and best way is to just use null/undefined as initialisation. I just copied this example from elsewhere in the codebase though, haha.